### PR TITLE
added saving / loading file by url

### PR DESCRIPTION
**Pull request by mmarganski**

### CODE REVIEW

File: src/lib/entities/message.entity.ts [hunk 1]
```diff
@@ -6,10 +6,10 @@ export class MessageEntity {
     @PrimaryGeneratedColumn()
     messageId: number
 
-    @Column()
+    @Column({length: 4000})
```
Comment (WadasGrzegorz) [inline src/lib/entities/message.entity.ts]: use data type text without length

File: src/modules/app/app.controller.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import { Controller, Get, Param, Res } from '@nestjs/common'
+import { fromBuffer } from 'file-type'
+import { Response } from 'express'
+const fs = require('fs')
```
Comment (WadasGrzegorz) [inline src/modules/app/app.controller.ts:5]: enter

File: src/modules/app/app.controller.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    async findImage(@Param('name') name: string, @Res() response: Response) {
+        try {
+            const path = `${process.env.IMAGES_PATH}\\${name}`
+            const image = fs.readFileSync(path)
+            const fileType = await fromBuffer(image)
+
+            response
+                .set('Content-Type', fileType?.mime)
+                .send(image)
+        } catch(e) {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.controller.ts]: remove `(e)` if you don't use it

File: src/modules/app/app.gateway.ts [hunk 4]
```diff
@@ -6,91 +6,132 @@ import {
     WsResponse
 } from '@nestjs/websockets'
 import { Server, Socket } from 'socket.io'
-import { Message, MessageCall, Rooms } from 'types'
+import { Message, MessageCall, MessageQueryResult } from 'lib/types/common'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: .

File: src/modules/app/app.gateway.ts [hunk 5]
```diff
@@ -6,91 +6,132 @@ import {
...
 import { AppService } from './app.service'
 
+import * as fs from 'fs'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: import order

File: src/modules/app/app.gateway.ts [hunk 6]
```diff
@@ -6,91 +6,132 @@ import {
...
+
+type CreateChatMessage = {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: move type to `./types`

File: src/modules/app/app.gateway.ts [hunk 7]
```diff
@@ -6,91 +6,132 @@ import {
...
-    activeUsers: Set<string> = new Set()
 
     constructor(private readonly appsService: AppService) {}
 
     @SubscribeMessage('getUsersList')
-    getUserList(client: Socket) {
-        client.emit('sendUsersList', [...this.users.entries()])
-        client.emit('activeUsers', [...this.activeUsers.keys()])
+    async getUserList(client: Socket) {
+        const users = await this.appsService.getUsersList()
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: active users and users should be in `Promise.all`

File: src/modules/app/app.gateway.ts [hunk 8]
```diff
@@ -6,91 +6,132 @@ import {
...
     }
 
     @SubscribeMessage('newRoom')
-    newRoom(client: Socket, roomName: string): WsResponse {
-        if (this.rooms.has(roomName)) {
+    async newRoom(client: Socket, roomName: string): Promise<WsResponse> {
+        const rooms = await this.appsService.getRoomNames()
+        const roomNames = rooms.map(({ roomName }) => roomName)
+
+        if(roomNames.includes(roomName)){
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: spaces

File: src/modules/app/app.gateway.ts [hunk 9]
```diff
@@ -6,91 +6,132 @@ import {
...
-            client.join(room)
-            this.rooms.set(room, [...this.rooms.get(room), client.id])
-        }
-
-        return { event: 'messageHistory', data: this.messages.get(room) }
+    async joinRoom(client: Socket, roomName: string): Promise<WsResponse> {
+        await this.appsService.addUserToRoom(roomName, client.id)
+        client.join(roomName)
+
+        const messageHistory = await  this.appsService.getMessagesByRoomName(roomName)
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: .

File: src/modules/app/app.gateway.ts [hunk 10]
```diff
@@ -6,91 +6,132 @@ import {
...
+        const formattedHistory = messageHistory
+            .map(({
+                messageText,
+                socketId,
+                ... others
+            }: MessageQueryResult) => ({
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: messageHistory should have a type

File: src/modules/app/app.gateway.ts [hunk 11]
```diff
@@ -6,91 +6,132 @@ import {
...
+    async joinPrivateRoom(client: Socket, userId: string): Promise<WsResponse> {
         const roomName = this.getPrivateRoomName(userId, client.id)
+        const room = await this.appsService.getRoomByName(roomName)
 
-        if (!this.rooms.has(roomName)) {
-            this.privateRooms.add(roomName)
-            this.rooms.set(roomName, [client.id, userId])
-            this.messages.set(roomName, [])
+        if (!room) {
+            await this.appsService.createRoom(roomName, true)
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts:96]: `Promise.all`

File: src/modules/app/app.gateway.ts [hunk 12]
```diff
@@ -6,91 +6,132 @@ import {
...
+                ...others
+            }))
+
+        return { event: 'messageHistory', data: formattedHistory }
     }
 
     @SubscribeMessage('sendMessage')
-    sendMessage(client: Socket, messageCall: MessageCall) {
+    async sendMessage(client: Socket, messageCall: MessageCall) {
+        const imagePath = messageCall.image === ''
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: `!messageCall.image`

File: src/modules/app/app.gateway.ts [hunk 13]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+        const fileType = metaData
+            .replace(/.*\//, '')
+            .replace(/;.*/, '')
+
+        return [buff, fileType]
+    }
+
+    saveImage(image: string, clientId: string): string{
+        const [content, fileType] = this.imageFromBase(image)
+        const imagePath = `${ process.env.IMAGES_PATH }\\${ clientId }-${ Math.random()
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: use uuid

File: src/modules/app/app.gateway.ts [hunk 14]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+            .toString(36)
+            .substr(2, 16) }.${ fileType }`
+            .trim()
+
+        fs.writeFile(`${ imagePath }`, content, () => {})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: use writeFileSync without empty callback

File: src/modules/app/app.gateway.ts [hunk 15]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
-        const time = new Date().getTime()
-        const newMessage: Message = {
-            userName: currentUserName,
-            userId: clientId,
-            avatar: userAvatar,
-            message: messageCall.messageContent,
-            image: messageCall.image,
-            date: time
+    async saveUser(socketId: string, avatar: string, userName: string){
+        if(avatar.includes('http://localhost')) {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: in this case You can only have one avatar for every users. Pleas also remove this avatar from git

File: src/modules/app/app.gateway.ts [hunk 16]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+            const newAvatarPath = 'http://localhost:3002/images/avatar.png'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: create env `API_URL`

File: src/modules/app/app.gateway.ts [hunk 17]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+
+            return this.appsService.createUser(socketId, userName, newAvatarPath)
         }
 
-        return newMessage
+        const [content, fileType] = this.imageFromBase(avatar)
+        const avatarPath = `${ process.env.IMAGES_PATH }\\${ socketId }.${ fileType }`
+        const avatarLocalPath = `${ process.env.IMAGES_LOCAL_PATH }\\${ socketId }.${ fileType }`
+
+        fs.writeFile(avatarPath, content, () => {})
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: aa

File: src/modules/app/app.gateway.ts [hunk 18]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+
+        return this.appsService.createUser(socketId, userName, avatarLocalPath)
     }
 
     getPrivateRoomName(userId: string, clientId: string) {
         return userId < clientId
-            ? `${userId} & ${clientId}`
-            : `${clientId} & ${userId}`
+            ? `${ userId } & ${ clientId }`
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: revert

File: src/modules/app/app.gateway.ts [hunk 19]
```diff
@@ -101,34 +142,61 @@ export class AppGateway implements OnGatewayDisconnect {
...
+            : `${ clientId } & ${ userId }`
     }
 
     getOtherPrivateRoomUser(roomName: string, clientId: string) {
-        const userId: string = roomName.replace(`${clientId}`, '')
+        const userId: string = roomName
+            .replace(`${ clientId }`, '')
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: remove spaces from string

File: src/modules/app/app.service.ts [hunk 20]
```diff
@@ -1,7 +1,7 @@
 import { InjectRepository } from '@nestjs/typeorm'
 import { Injectable } from '@nestjs/common'
 import { Repository } from 'typeorm'
-import { CreateChatMessage, Message } from 'lib/types/common'
+import { CreateChatMessage, Message, MessageQueryResult } from 'lib/types/common'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: .

File: src/modules/app/app.service.ts [hunk 21]
```diff
@@ -130,9 +143,17 @@ export class AppService {
...
         }
 
         return newMessage
     }
+
+    private async createUserRoom (socketId: string, roomName: string) {
+        return this.userRoomRepository.save({
+            roomName,
+            socketId,
+            createdAt: new Date(),
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: this fields should have default value

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: package.json [hunk 22]
```diff
@@ -45,6 +46,7 @@
     "@types/jest": "26.0.24",
     "@types/node": "16.0.0",
     "@types/supertest": "2.0.11",
+    "@types/uuid": "^8.3.1",
```
Comment (WadasGrzegorz) [inline package.json]: .

File: src/lib/entities/user-room.entity.ts [hunk 23]
```diff
@@ -12,9 +12,9 @@ export class UserRoomEntity {
     @Column()
     roomName: string
 
-    @Column()
-    createdAt: Date
+    @Column('timestamp', { default: () => 'CURRENT_TIMESTAMP'})
```
Comment (WadasGrzegorz) [inline src/lib/entities/user-room.entity.ts]: you can use 
```    
@CreateDateColumn()
    createdAt: Date

    @UpdateDateColumn()
    updatedAt: Date
```

File: src/modules/app/app.gateway.ts [hunk 24]
```diff
@@ -5,18 +5,12 @@ import {
...
 import * as fs from 'fs'
-
-type CreateChatMessage = {
-    messageText: string,
-    image: string,
-    socketId: string,
-    roomName: string
-}
+import { v4 as uuidv4 } from 'uuid'
+import { AppService } from './app.service'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: import order

File: src/modules/app/app.gateway.ts [hunk 25]
```diff
@@ -5,18 +5,12 @@ import {
...
+import { Server, Socket } from 'socket.io'
+import { Message, MessageCall, CreateChatMessage } from 'lib/types/common'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: use index

File: src/modules/app/app.gateway.ts [hunk 26]
```diff
@@ -5,18 +5,12 @@ import {
...
+import { MessageQueryResult } from './dao/queryResults.dao'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: use index

File: src/modules/app/app.gateway.ts [hunk 27]
```diff
@@ -58,7 +54,7 @@ export class AppGateway implements OnGatewayDisconnect {
         const rooms = await this.appsService.getRoomNames()
         const roomNames = rooms.map(({ roomName }) => roomName)
 
-        if(roomNames.includes(roomName)){
+        if(roomNames.includes(roomName)) {
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts]: space after `if`

File: src/modules/app/app.service.ts [hunk 28]
```diff
@@ -1,8 +1,9 @@
 import { InjectRepository } from '@nestjs/typeorm'
 import { Injectable } from '@nestjs/common'
 import { Repository } from 'typeorm'
-import { CreateChatMessage, Message, MessageQueryResult } from 'lib/types/common'
+import { CreateChatMessage, Message } from 'lib/types/common'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

File: src/modules/app/app.service.ts [hunk 29]
```diff
@@ -1,8 +1,9 @@
...
 import { UserEntity, RoomEntity, MessageEntity, UserRoomEntity } from 'lib/entities'
+import { MessageQueryResult } from './dao/queryResults.dao'
```
Comment (WadasGrzegorz) [inline src/modules/app/app.service.ts]: aa

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/app/app.gateway.ts [hunk 30]
```diff
@@ -5,92 +5,134 @@ import {
...
+
+        if (roomNames.includes(roomName)) {
             return null
         }
 
-        const [currentUserName] = this.users.get(client.id)
-
-        this.rooms.set(roomName, [currentUserName])
-        this.messages.set(roomName, [])
+        await this.appsService.createRoom(roomName, false)
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts:61]: Promise.all ?

File: src/modules/app/app.gateway.ts [hunk 31]
```diff
@@ -128,7 +193,8 @@ export class AppGateway implements OnGatewayDisconnect {
     }
 
     getOtherPrivateRoomUser(roomName: string, clientId: string) {
-        const userId: string = roomName.replace(`${clientId}`, '')
+        const userId: string = roomName
```
Comment (WadasGrzegorz) [inline src/modules/app/app.gateway.ts:196]: You don't need to type it replace always return string

**Review by WadasGrzegorz — CHANGES_REQUESTED** [review]

File: src/modules/app/app.gateway.ts (hunk 30 above)
Comment (mmarganski) [reply src/modules/app/app.gateway.ts:61]: to work properly appsService.addUserToRoom() requires completed appsService.createRoom() , thus Promise.all in this case might be invalid

Comment (WadasGrzegorz) [reply src/modules/app/app.gateway.ts:61]: :+1:

**Review by WadasGrzegorz — APPROVED** [review]

**Review by jpudysz — COMMENTED** [review]
Good job!

**Pull request merged**